### PR TITLE
developers_guides: add user guide for keyword detection

### PR DESCRIPTION
**Pull request by keyonjie**
Add Keyword Detection driver implementation and user guide to the third
party features user guide of the SOF linux driver documentation.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+   # PCM  <-----(pipe 8)--------- host <---------------- KPB <------- Volume <--- DMIC (dmic16k)
+   #  |                                                   |
+   #  |                                                   |
+   #  |                                                   |
+   # Detector Sink <---Detector(pipe 9) <--- selector <---+
+
+We are using DAPM events to trigger the detect pipeline(pipe 9). Here the ‘Detector Sink’ is a virtual DAPM widget(visible to the driver, but not to the FW), it is used to send pipeline control IPCs(hw_params, trigger start/stop, hw_free) to firmware. These control IPCs are sent to FW in the sequence like this (1->2(2.1->2.2)->3->4->5->6(6.1->6.2)):
+
+.. code-block:: none
+
```
Comment (dbaluta) [inline developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:24]: Can we use rst table format? https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html#simple-tables

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst [hunk 2]
```diff
@@ -0,0 +1,140 @@
+.. _sof_driver_arch:
+
+Keyword Detection Driver implementation and user guide
+######################################################
+
+.. contents::
```
Comment (dbaluta) [inline developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:6]: Can you add a small introduction of what keyword detection is and how it is supposed to work? Also it is not very clear how the detection algorithm is loaded and used by SOF.

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst (hunk 1 above)
Comment (keyonjie) [reply developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:24]: Just tried and looks the simple table doesn't work, maybe we are lack of numpydoc or autodoc extensions, but switching to use csv-table works fine, updating it now.

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst (hunk 2 above)
Comment (keyonjie) [reply developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:6]: Sure, done now.

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+KPB kcontrol
+------------
+
+The kcontrol for KPB configuration is(amixer controls | grep “KPB”):
+
+.. code-block:: none
+
+   numid=13,iface=MIXER,name='KPBM8.0 KPB'
+
+The initial value of it is in KPB_priv section from sof/tools/topology/sof/pipe-kfbm-capture.m4(The first 32 Bytes are abi header), and it is aligned with the definition of struct sof_kpb_config in sof/include/user/kpb.h.
```
Comment (dbaluta) [inline developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:67]: Small nits: add a space between a letter and an open parenthesis. here ...``m4 (the first 32 bytes). Please use lower letter for 'The' and 'Bytes'). Same issue below.

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+-----------------
+
+The kcontrol for selector configuration is(amixer controls | grep “SELECTOR”):
+
+.. code-block:: none
+
+   numid=16,iface=MIXER,name='SELECTOR9.0 SELECTOR'
+
+The initial value of it is in SELECTOR_priv section from sof/tools/topology/sof/pipe-detect.m4(The first 32 Bytes are abi header), and it is aligned with the definition of struct sof_sel_config in sof/include/user/selector.h.
+
```
Comment (dbaluta) [inline developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:79]: Same here ->`` .m4<space>(the first 32 bytes)``

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+---------------------------------------------
+
+The kcontrol for detector configuration is(amixer controls | grep “Detector Config”):
+
+.. code-block:: none
+
+   numid=14,iface=MIXER,name='DETECT9.0 Detector Config'
+
+The initial value of it is in DETECTOR_priv section from sof/tools/topology/m4/detect_test_coef.m4(The first 32 Bytes are abi header), and it is aligned with the definition of struct sof_detect_test_config in sof/include/user/detect_test.h.
+
```
Comment (dbaluta) [inline developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:90]: Ditto!

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+   numid=15,iface=MIXER,name='DETECT9.0 Hotword Model'
+
+This is vendor specific, by default, it will be initialized to 64 Bytes 0s only.
+
+sof-ctl tool
+************
+
+For all those TLV Bytes kcontrols, after pipeline/PCM created, we can use the sof tool named sof-ctl(source located in sof/tools/ctl/ctl.c, run “./scripts/build-tools.sh” in sof folder to build and generate it) to configure/update with new blob, e.g.
+
+To set,
```
Comment (dbaluta) [inline developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:107]: Same here ```sof-ctl<space>(source ..)

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+   #arecord -Dhw:0,8 -M -N -c 2 -f S16_LE -r 16000 --buffer-size=64000 tmp.wav -vvv
+
+In another terminal, run:
+
+.. code-block:: none
+
+   #echo freeze > /sys/power/state
+
+Then the Keyword Detection feature will be activated at S0Ix, say keyword to trigger the Keyword detected, the system will be woken up and the keyword and command data will be captured.
+
```
Comment (dbaluta) [inline developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:148]: No need for empty new line at the end.

**Comment by dbaluta** [comment]
@keyonjie Thank you! This looks super good. So, aside from the small nits on my side this commit looks good to me.

File: developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst (hunk 3 above)
Comment (keyonjie) [reply developer_guides/linux_driver/third_party/keyword_detect/keyword_detect.rst:67]: Thanks.

**Comment by keyonjie** [comment]
Thank you @dbaluta, updated now.

**Comment by dbaluta** [comment]
@keyonjie this is good to merge. 

I have one question since we are here: How is the algorithm for keyword detection loaded on DSP? Is there such a dummy algorithm to test this feature?

**Comment by keyonjie** [comment]
> @keyonjie this is good to merge.
> 
> I have one question since we are here: How is the algorithm for keyword detection loaded on DSP? Is there such a dummy algorithm to test this feature?

By default we already have a dummy algorithm to test it, so the algorithm is loaded together with the base FW with today's implementation.

**Comment by keyonjie** [comment]
@deb-intel Thanks.

BTW, why it doesn't reflect on https://thesofproject.github.io/latest/developer_guides/linux_driver/index.html yet, do we need trigger a rebuild for it?

**Pull request merged**